test/framework/service.go

```diff
@@ -52,6 +53,41 @@ func (f *Framework) CreateServiceAndWaitUntilReady(ctx context.Context, namespac
 	return finalizerFn, nil
 }
 
+func (f *Framework) CreateOrUpdateServiceAndWaitUntilReady(ctx context.Context, namespace string, service *v1.Service) (FinalizerFn, error) {
+	finalizerFn := func() error { return f.DeleteServiceAndWaitUntilGone(ctx, namespace, service.Name) }
+
+	s, err := f.KubeClient.CoreV1().Services(namespace).Get(ctx, service.Name, metav1.GetOptions{})
+	if err != nil && !apierrors.IsNotFound(err) {
+		return finalizerFn, errors.Wrap(err, fmt.Sprintf("getting service %v failed", service.Name))
+	}
+
+	if apierrors.IsNotFound(err) {
+		// Service doesn't exists -> Create
+		if _, err := f.KubeClient.CoreV1().Services(namespace).Create(ctx, service, metav1.CreateOptions{}); err != nil {
+			return finalizerFn, errors.Wrap(err, fmt.Sprintf("creating service %v failed", service.Name))
+		}
+	} else {
+		// must set these immutable fields from the existing service to prevent update fail
+		service.ObjectMeta.ResourceVersion = s.ObjectMeta.ResourceVersion
+		service.Spec.ClusterIP = s.Spec.ClusterIP
+		service.Spec.ClusterIPs = s.Spec.ClusterIPs
+		service.Spec.IPFamilies = s.Spec.IPFamilies
+
+		// Service already exists -> Update
+		if _, err := f.KubeClient.CoreV1().Services(namespace).Update(ctx, service, metav1.UpdateOptions{}); err != nil {
+			return finalizerFn, errors.Wrap(err, fmt.Sprintf("updating service %v failed", service.Name))
+		}
+
+		// Allow controller to update the endpoints
+		time.Sleep(15 * time.Second)
```

Comment:
do we really need this sleep? I'd expect `WaitForServiceReady` to be eventual?

Reply:
Sorry for the consistent lag on reviewing this PR and thanks for continuing to see it through. 

Would checking the revision version in such cases be useful to determine if there has been an update instead of the sleep or perhaps some other mechanism? Regardless, I think in the context of this being a test function we can and should remove it unless it begins to flake!

Reply:
@PhilipGough agree. Done, please have a look. :D